### PR TITLE
New task flow buttons and actions [SCI-4824] [SCI-4833]

### PR DESCRIPTION
**Pull request by urbanrotnik**
Jira ticket: [SCI-4824](https://biosistemika.atlassian.net/browse/SCI-4824)
Jira ticket: [SCI-4833](https://biosistemika.atlassian.net/browse/SCI-4833)

### What was done
- Basic button with dropdown
- Controller action
- JS call 
- Clean old code

### Note:
Not finished, will be when we have DB layer finished

### CODE REVIEW

File: app/permissions/my_module.rb [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+    user.is_technician_or_higher_of_project?(my_module.experiment.project)
+  end
+
+  # module: assign/unassign sample
+  # NOTE: Use 'module_page? &&' before calling this permission!
+  can :assign_sample_to_module do |user, my_module|
+    user.is_technician_or_higher_of_project?(my_module.experiment.project)
+  end
+
+  # module: complete/uncomplete
```
Comment (mlorb) [inline app/permissions/my_module.rb:51]: Fix also comment.

File: app/controllers/my_modules_controller.rb [hunk 2]
```diff
@@ -249,109 +249,24 @@ def results
...
-          format.json do
-            render json: {
-              new_btn: render_to_string(partial: new_btn_partial),
-              completed: completed,
-              module_header_due_date: render_to_string(
-                partial: 'my_modules/module_header_due_date.html.erb',
-                locals: { my_module: @my_module }
-              ),
-              module_state_label: render_to_string(
-                partial: 'my_modules/module_state_label.html.erb',
```
Comment (mlorb) [inline app/controllers/my_modules_controller.rb:283]: Do we still need partial `my_modules/module_state_label.html.erb`? Probably you can remove it.

File: app/controllers/my_modules_controller.rb [hunk 3]
```diff
@@ -249,109 +249,24 @@ def results
...
 
-  def task_completion_activity
-    completed = @my_module.completed?
-    log_activity(completed ? :complete_task : :uncomplete_task)
-    start_work_on_next_task_notification
-  end
-
-  def start_work_on_next_task_notification
-    if @my_module.completed?
-      title = t('notifications.start_work_on_next_task',
```
Comment (mlorb) [inline app/controllers/my_modules_controller.rb:332]: You should remove locales.

File: app/controllers/my_modules_controller.rb [hunk 4]
```diff
@@ -249,109 +249,24 @@ def results
...
-                user: current_user.full_name,
-                module: @my_module.name)
-      message = t('notifications.start_work_on_next_task_message',
```
Comment (mlorb) [inline app/controllers/my_modules_controller.rb:335]: You should remove locales.

File: app/assets/javascripts/my_modules.js [hunk 5]
```diff
@@ -226,34 +226,37 @@ function bindEditTagsAjax() {
...
-          initDueDatePicker();
-          $('.task-state-label').html(data.module_state_label);
-          button.find('button').replaceWith(data.new_btn);
-        },
-        error: function() {
+function applyTaskStatusChangeCallBack() {
+  $('#dropdownTaskFlowList > li[data-state-id]').on('click', function() {
+    var list = $('#dropdownTaskFlowList');
+    var item = $(this);
+    var conteiner = list.closest('.task-flows');
```
Comment (mlorb) [inline app/assets/javascripts/my_modules.js:233]: Typo.

File: app/assets/javascripts/my_modules.js [hunk 6]
```diff
@@ -226,34 +226,37 @@ function bindEditTagsAjax() {
...
+    console.log('click')
```
Comment (mlorb) [inline app/assets/javascripts/my_modules.js:234]: console.log

File: app/assets/javascripts/my_modules.js [hunk 7]
```diff
@@ -226,34 +226,37 @@ function bindEditTagsAjax() {
...
+    $.ajax({
+      url: list.data('link-url'),
+      type: 'PATCH',
+      contentType: 'application/json',
+      data: JSON.stringify({ status_id: item.data('state-id') }),
+      success: function(data) {
+        console.log(data.content);
```
Comment (mlorb) [inline app/assets/javascripts/my_modules.js:241]: console.log

File: app/assets/javascripts/my_modules.js [hunk 8]
```diff
@@ -226,34 +226,37 @@ function bindEditTagsAjax() {
...
+        conteiner.html(data.content)
+        HelperModule.flashAlertMsg('Greate status will be updated', 'success');
```
Comment (mlorb) [inline app/assets/javascripts/my_modules.js:243]: Put it in locales.

Comment (urbanrotnik) [reply app/assets/javascripts/my_modules.js:243]: > Nice job 🎉
> I have some comments.

Its WIP, not finished yet, but anyway, thanks for comments ;)

**Comment by urbanrotnik** [comment]
> Nice job 🎉
> I have some comments.
Its WIP, not finished yet, but anyway, thanks for comments ;)

File: app/views/my_modules/_state_buttons.html.erb [hunk 9]
```diff
@@ -1,16 +0,0 @@
-<div class="pull-right my_module-state-buttons">
```
[the remark this answers is not in the document]
Comment (urbanrotnik) [reply app/views/my_modules/_state_buttons.html.erb:1]: I think this should be solved in another task? Where we decouple el. signatures from this flow?

[the remark this answers is not in the document]
Comment (urbanrotnik) [reply app/views/my_modules/_state_buttons.html.erb:1]: https://biosistemika.atlassian.net/browse/SCI-4857

File: app/views/my_modules/status_flow/_task_flow_button.html.erb [hunk 10]
```diff
@@ -0,0 +1,31 @@
+<% status = my_module.my_module_status %>
+<div class="status-label">Status</div>
```
Comment (okriuchykhin) [inline app/views/my_modules/status_flow/_task_flow_button.html.erb:2]: What should be here?

File: app/views/my_modules/status_flow/_task_flow_button.html.erb [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+          aria-expanded="true"
+          style="<%= "background-color: ##{status.color}" %>;">
+    <span><%= status.name %></span>
+    <span class="caret pull-right"></span>
+  </button>
+  <ul class="dropdown-menu" aria-labelledby="dropdownTaskFlow" id="dropdownTaskFlowList" data-link-url="<%= update_state_my_module_url(my_module) %>">
+    <% unless status.initial_status? %>
+      <% previous_s = status.previous_status %>
+      <li data-state-id="<%= previous_s.id %>">
+        <span>Return to -></span> <span style="<%= "background-color: ##{previous_s.color}" %>"><%= previous_s.name %></span>
```
Comment (okriuchykhin) [inline app/views/my_modules/status_flow/_task_flow_button.html.erb:17]: I18n?

File: app/views/my_modules/status_flow/_task_flow_button.html.erb (hunk 10 above)
Comment (urbanrotnik) [reply app/views/my_modules/status_flow/_task_flow_button.html.erb:2]: There should be I18n :D

File: app/views/my_modules/status_flow/_task_flow_button.html.erb [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+      </li>
+    <% end %>
+
+    <% unless status.final_status? %>
+      <% next_s = status.next_status %>
+      <li data-state-id="<%= next_s.id %>">
+        <span>Move to -></span> <span style="<%= "background-color: ##{next_s.color}" %>"><%= next_s.name %></span>
```
Comment (okriuchykhin) [inline app/views/my_modules/status_flow/_task_flow_button.html.erb:24]: I18n?

File: app/views/my_modules/status_flow/_task_flow_button.html.erb [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+      </li>
+    <% end %>
+    <li>
+      <span><a href="#">View task flow</a></span>
```
Comment (okriuchykhin) [inline app/views/my_modules/status_flow/_task_flow_button.html.erb:28]: I18n?

File: app/assets/javascripts/my_modules.js [hunk 14]
```diff
@@ -226,34 +226,33 @@ function bindEditTagsAjax() {
...
+  $('.task-flows').on('click', '#dropdownTaskFlowList > li[data-state-id]', function() {
+    var list = $('#dropdownTaskFlowList');
+    var item = $(this);
+    var container = list.closest('.task-flows');
+    animateSpinner();
+    $.ajax({
+      url: list.data('link-url'),
+      type: 'PATCH',
+      contentType: 'application/json',
+      data: JSON.stringify({ status_id: item.data('state-id') }),
```
Comment (okriuchykhin) [inline app/assets/javascripts/my_modules.js:239]: Why JSON.stringify is needed here?

**Pull request merged**